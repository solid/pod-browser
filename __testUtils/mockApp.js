/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  createThing,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { vcard, rdf, schema } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { vcardExtras } from "../src/addressBook";

export const oidcPrefix = "http://www.w3.org/ns/solid/oidc#";
export const TOS_PREDICATE = `${oidcPrefix}tos_uri`;
export const POLICY_PREDICATE = `${oidcPrefix}policy_uri`;
export const CONTACTS_PREDICATE = `${oidcPrefix}contacts`;
export const LOGO_PREDICATE = `${oidcPrefix}logo_uri`;

export const APP_WEBID = "https://mockappurl.com";
export const APP_POLICY_URL = "https://mockappurl.com/privacy-policy";
export const APP_TOS_URL = "https://mockappurl.com/terms-of-service";

const contactThing = chain(
  createThing({ url: "https://mockappurl.com#contacts" }),
  (t) => addStringNoLocale(t, vcard.fn, "Example Contact"),
  (t) => addStringNoLocale(t, vcard.hasTelephone, "555-5555"),
  (t) => addStringNoLocale(t, vcard.hasEmail, "example@email.com"),
  (t) => addUrl(t, vcard.url, "https://examplewebsite.com")
);

const dataset = setThing(createSolidDataset(), contactThing);

export function mockApp() {
  return chain(
    createThing({ url: APP_WEBID }),
    (t) => addStringNoLocale(t, vcard.fn, "Mock App"),
    (t) => addUrl(t, vcardExtras("WebId"), APP_WEBID),
    (t) => addUrl(t, rdf.type, schema.SoftwareApplication), // just guessing what the type might be for apps
    (t) => addUrl(t, TOS_PREDICATE, APP_TOS_URL),
    (t) => addUrl(t, POLICY_PREDICATE, APP_POLICY_URL),
    (t) => addUrl(t, CONTACTS_PREDICATE, "https://mockappurl.com#contacts")
  );
}

export function mockAppDataset() {
  return setThing(dataset, mockApp());
}
