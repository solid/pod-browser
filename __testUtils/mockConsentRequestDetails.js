/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

async function getConsentRequestDetails() {
  return {
    "@context": [
      "https://www.w3.org/2018/credentials/v1",
      "https://w3id.org/security/suites/ed25519-2020/v1",
      "https://consent.pod.inrupt.com/credentials/v1",
    ],
    id: "https://consent.pod.inrupt.com/vc/53973727-f4d0-9e8dbdc041fd",
    type: ["VerifiableCredential", "SolidCredential", "SolidConsentRequest"],
    issuer: "https://consent.pod.inrupt.com",
    issuanceDate: "2021-05-26T16:40:03Z",
    expirationDate: "2021-06-23T16:40:03Z",
    credentialSubject: {
      id: "https://pod.inrupt.com/bob/profile/card#me",
      inbox: "https://pod.inrupt.com/bob/inbox/",
      hasConsent: {
        mode: ["Read", "Write"],
        hasStatus: "ConsentStatusRequested",
        forPersonalData: "https://pod.inrupt.com/alice/private/data",
        forPurpose: "https://example.com/SomeSpecificPurpose",
      },
    },
    proof: {
      created: "2021-05-26T16:40:03.009Z",
      proofPurpose: "assertionMethod",
      proofValue: "eqp8h_kL1DwJCpn65z-d1Arnysx6b11...jb8j0MxUCc1uDQ",
      type: "Ed25519Signature2020",
      verificationMethod: "https://consent.pod.inrupt.com/key/396f686b4ec",
    },
  };
}

export default getConsentRequestDetails;
