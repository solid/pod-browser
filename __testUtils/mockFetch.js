/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

export default function mockFetch(responses = {}) {
  return jest.fn().mockImplementation((url, init) => {
    const responseFn = responses[url.toString()];
    const responseObject = responseFn;
    const method = init?.method || "GET";
    let response;
    if (responseFn && typeof responseObject[method] !== "undefined") {
      response = responseObject[method](url, init);
    } else if (responseFn) {
      response = responseFn(url, init);
    } else {
      throw new Error(`URL (${url}) not mocked properly`);
    }
    Object.defineProperty(response, "url", { value: url });
    return Promise.resolve(response);
  });
}

export function mockSolidClientRequest(body, method = "GET", headers = {}) {
  return new Request(body, {
    headers: new Headers(headers),
    method,
  });
}

export function mockTurtleResponse(status, body = "") {
  return new Response(body, {
    status,
    headers: {
      "Content-Type": "text/turtle",
    },
  });
}
