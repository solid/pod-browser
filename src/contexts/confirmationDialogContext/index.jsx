/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { createContext, useState } from "react";
import T from "prop-types";

export const defaultConfirmationDialogContext = {
  confirmed: false,
  content: null,
  open: false,
  setConfirmed: () => {},
  setContent: () => {},
  setOpen: () => {},
  setTitle: () => {},
  title: "Confirmation",
  cancelText: "Cancel",
  confirmText: "Confirm",
  setCancelText: () => {},
  setConfirmText: () => {},
  closeDialog: () => {},
  omitCancelButton: false,
};
const ConfirmationDialogContext = createContext(
  defaultConfirmationDialogContext
);

export default ConfirmationDialogContext;

function ConfirmationDialogProvider({ children }) {
  const [open, setOpen] = useState(null);
  const [content, setContent] = useState(null);
  const [title, setTitle] = useState("Confirmation");
  const [cancelText, setCancelText] = useState("Cancel");
  const [confirmText, setConfirmText] = useState("Confirm");
  const [confirmed, setConfirmed] = useState(null);
  const [omitCancelButton, setOmitCancelButton] = useState(false);

  const closeDialog = () => {
    setConfirmed(null);
    setOpen(null);
    setContent(null);
    setTitle(null);
  };

  return (
    <ConfirmationDialogContext.Provider
      value={{
        content,
        confirmed,
        open,
        setContent,
        setConfirmed,
        setOpen,
        setTitle,
        title,
        cancelText,
        setCancelText,
        confirmText,
        setConfirmText,
        closeDialog,
        omitCancelButton,
        setOmitCancelButton,
      }}
    >
      {children}
    </ConfirmationDialogContext.Provider>
  );
}

ConfirmationDialogProvider.propTypes = {
  children: T.node,
};

ConfirmationDialogProvider.defaultProps = {
  children: null,
};

export { ConfirmationDialogProvider };
