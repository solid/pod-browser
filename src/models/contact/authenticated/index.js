/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createThing,
  getStringNoLocale,
  getUrl,
  setThing,
} from "@inrupt/solid-client";
import { foaf } from "rdf-namespaces";
import { chain } from "../../../solidClientHelpers/utils";

const AGENT_PREDICATE = "http://www.w3.org/ns/solid/acp#agent";
export const AUTHENTICATED_AGENT_PREDICATE =
  "http://www.w3.org/ns/solid/acp#AuthenticatedAgent";

export const AUTHENTICATED_AGENT_NAME = "Anyone signed in";
export const AUTHENTICATED_AGENT_TYPE = "authenticated";

export const AUTHENTICATED_AGENT = {
  isOfType: (thing) =>
    thing && getUrl(thing, AGENT_PREDICATE) === AUTHENTICATED_AGENT_PREDICATE,
  getName: (thing) => getStringNoLocale(thing, foaf.name),
  getAvatarProps: () => ({
    icon: "user-lock",
    src: null,
    variant: "authenticated",
  }),
};

export const createAuthenticatedAgent = (dataset) => {
  const thing = chain(
    createThing(),
    (t) =>
      addUrl(
        t,
        "http://www.w3.org/ns/solid/acp#agent",
        "http://www.w3.org/ns/solid/acp#AuthenticatedAgent"
      ),
    (t) => addStringNoLocale(t, foaf.name, "Anyone signed in")
  );
  const updatedDataset = setThing(dataset, thing);
  return {
    thing,
    dataset: updatedDataset,
    type: AUTHENTICATED_AGENT,
  };
};
