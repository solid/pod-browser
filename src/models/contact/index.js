/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { getThingAll } from "@inrupt/solid-client";
import { getContactIndex } from "./collection";

/*
 * Contacts represent the agents or groups in a user's AddressBook
 */

/* Model functions */
export function getContactAllFromContactsIndex(contactIndex) {
  const { dataset, type } = contactIndex;
  return getThingAll(dataset)
    .filter((contact) => type.isOfType(contact))
    .map((thing) => {
      return {
        thing,
        dataset,
      };
    });
}

export function getContactAllFromContactIndexArray(contactIndexArray) {
  return contactIndexArray
    .map((index) => getContactAllFromContactsIndex(index))
    .reduce((memo, contacts) => memo.concat(contacts), []);
}

export async function getContactAll(addressBook, types, fetch) {
  const contactIndexArray = await Promise.all(
    types.map(async (type) => getContactIndex(addressBook, type, fetch))
  );
  return getContactAllFromContactIndexArray(contactIndexArray);
}
