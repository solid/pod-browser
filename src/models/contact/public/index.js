/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createThing,
  getStringNoLocale,
  getUrl,
  setThing,
} from "@inrupt/solid-client";
import { foaf } from "rdf-namespaces";
import { chain } from "../../../solidClientHelpers/utils";

const AGENT_PREDICATE = "http://www.w3.org/ns/solid/acp#agent";
export const PUBLIC_AGENT_NAME = "Anyone";
export const PUBLIC_AGENT_TYPE = "public";

export const PUBLIC_AGENT_PREDICATE =
  "http://www.w3.org/ns/solid/acp#PublicAgent";

export const PUBLIC_AGENT = {
  isOfType: (thing) =>
    thing && getUrl(thing, AGENT_PREDICATE) === PUBLIC_AGENT_PREDICATE,
  getName: (thing) => getStringNoLocale(thing, foaf.name),
  getAvatarProps: () => ({
    icon: "globe",
    src: null,
    variant: "public",
  }),
};

export const createPublicAgent = (dataset) => {
  const thing = chain(
    createThing(),
    (t) =>
      addUrl(
        t,
        "http://www.w3.org/ns/solid/acp#agent",
        "http://www.w3.org/ns/solid/acp#PublicAgent"
      ),
    (t) => addStringNoLocale(t, foaf.name, "Anyone")
  );
  const updatedDataset = setThing(dataset, thing);
  return {
    thing,
    dataset: updatedDataset,
    type: PUBLIC_AGENT,
  };
};
