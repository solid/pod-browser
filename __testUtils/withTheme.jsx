/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { render } from "@testing-library/react";
import T from "prop-types";
import { StylesProvider, ThemeProvider } from "@inrupt/prism-react-components";
import { FeatureProvider } from "../src/contexts/featureFlagsContext";
import defaultTheme from "../src/theme";

export default function WithTheme(props) {
  const { theme, children } = props;
  return (
    <StylesProvider generateClassName={(rule) => `PodBrowser-${rule.key}`}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </StylesProvider>
  );
}

WithTheme.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  theme: T.object,
  children: T.node.isRequired,
};

WithTheme.defaultProps = {
  theme: defaultTheme,
};

export const renderWithTheme = (children, theme = defaultTheme) => {
  return render(
    <FeatureProvider>
      <WithTheme theme={theme}>{children}</WithTheme>
    </FeatureProvider>
  );
};
