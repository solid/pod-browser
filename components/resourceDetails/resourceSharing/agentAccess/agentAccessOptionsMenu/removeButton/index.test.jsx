/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { waitFor } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";
// eslint-disable-next-line camelcase
import { acp_v2 } from "@inrupt/solid-client";
import { renderWithTheme } from "../../../../../../__testUtils/withTheme";
import mockAccessControl from "../../../../../../__testUtils/mockAccessControl";
import RemoveButton, {
  handleRemovePermissions,
  TESTCAFE_ID_REMOVE_BUTTON,
} from "./index";
import { PUBLIC_AGENT_PREDICATE } from "../../../../../../src/models/contact/public";
import { AUTHENTICATED_AGENT_PREDICATE } from "../../../../../../src/models/contact/authenticated";
import { ConfirmationDialogProvider } from "../../../../../../src/contexts/confirmationDialogContext";
import {
  TESTCAFE_ID_CONFIRM_BUTTON,
  TESTCAFE_ID_CONFIRMATION_DIALOG,
  TESTCAFE_ID_CONFIRMATION_DIALOG_TITLE,
} from "../../../../../confirmationDialog";
import AccessControlContext from "../../../../../../src/contexts/accessControlContext";

const resourceIri = "/iri/";
const resourceUrl = "http://example.com/resource";
const webId = "https://example.com/profile/card#me";
const name = "Example Agent";
const profile = {
  avatar: null,
  name,
  webId,
};
const permission = { webId, alias: "editors" };

describe("AgentAccessOptionsMenu", () => {
  test("it renders a button which triggers the opening of the menu", () => {
    const { asFragment, getByTestId } = renderWithTheme(
      <RemoveButton
        resourceIri={resourceIri}
        permission={permission}
        profile={profile}
        setLoading={jest.fn()}
        setLocalAccess={jest.fn()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    const button = getByTestId("remove-button");
    expect(button).toBeDefined();
  });
  test("clicking on delete button renders a confirmation dialog with the correct title and message", async () => {
    const { getByTestId, findByTestId } = renderWithTheme(
      <ConfirmationDialogProvider>
        <RemoveButton
          resourceIri={resourceIri}
          permission={permission}
          profile={profile}
          setLoading={jest.fn()}
          setLocalAccess={jest.fn()}
        />
      </ConfirmationDialogProvider>
    );
    const button = getByTestId(TESTCAFE_ID_REMOVE_BUTTON);
    userEvent.click(button);
    const dialog = await findByTestId(TESTCAFE_ID_CONFIRMATION_DIALOG);
    expect(dialog).toBeInTheDocument();
    expect(
      getByTestId(TESTCAFE_ID_CONFIRMATION_DIALOG_TITLE)
    ).toHaveTextContent(`Remove ${name}'s access from iri`);
  });
  test("if no profile is available, confirmation dialog title is agent's webId", async () => {
    const { getByTestId } = renderWithTheme(
      <ConfirmationDialogProvider>
        <RemoveButton
          resourceIri={resourceIri}
          permission={permission}
          profile={null}
          setLoading={jest.fn()}
          setLocalAccess={jest.fn()}
        />
      </ConfirmationDialogProvider>
    );
    const button = getByTestId(TESTCAFE_ID_REMOVE_BUTTON);
    userEvent.click(button);
    expect(
      getByTestId(TESTCAFE_ID_CONFIRMATION_DIALOG_TITLE)
    ).toHaveTextContent(`Remove ${webId}'s access from iri`);
  });
  test("if dialog is confirmed, it removes agent from permissions", async () => {
    const accessControl = mockAccessControl();
    const { getByTestId } = renderWithTheme(
      <AccessControlContext.Provider value={{ accessControl }}>
        <ConfirmationDialogProvider>
          <RemoveButton
            resourceIri={resourceIri}
            permission={permission}
            profile={profile}
            setLoading={jest.fn()}
            setLocalAccess={jest.fn()}
          />
        </ConfirmationDialogProvider>
      </AccessControlContext.Provider>
    );
    const button = getByTestId(TESTCAFE_ID_REMOVE_BUTTON);
    userEvent.click(button);
    const confirmButton = getByTestId(TESTCAFE_ID_CONFIRM_BUTTON);
    userEvent.click(confirmButton);
    waitFor(() =>
      expect(accessControl.removeAgentFromPolicy).toHaveBeenCalledWith(
        webId,
        "editors"
      )
    );
  });
  test("if agent is public agent or authenticated agent, it does not render a dialog and removes it from permissions", async () => {
    const accessControl = mockAccessControl();
    const { getByTestId } = renderWithTheme(
      <AccessControlContext.Provider value={{ accessControl }}>
        <ConfirmationDialogProvider>
          <RemoveButton
            resourceIri={resourceIri}
            permission={{ webId: PUBLIC_AGENT_PREDICATE, alias: "editors" }}
            profile={null}
            setLoading={jest.fn()}
            setLocalAccess={jest.fn()}
          />
        </ConfirmationDialogProvider>
      </AccessControlContext.Provider>
    );
    const button = getByTestId(TESTCAFE_ID_REMOVE_BUTTON);
    userEvent.click(button);
    waitFor(() => {
      expect(
        getByTestId(TESTCAFE_ID_CONFIRMATION_DIALOG)
      ).not.toBeInTheDocument();
      expect(accessControl.setRulePublic).toHaveBeenCalledWith(
        "editors",
        false
      );
    });
  });
});

describe("handleRemovePermissions", () => {
  const setLoading = jest.fn();
  const accessControl = mockAccessControl();
  const mutateResourceInfo = jest.fn();
  const setLocalAccess = jest.fn();
  const policyName = "editors";
  const acr = acp_v2.mockAcrFor(resourceUrl);

  const handler = handleRemovePermissions({
    setLoading,
    accessControl,
    mutateResourceInfo,
    setLocalAccess,
  });

  beforeEach(() => {
    accessControl.removeAgentFromPolicy.mockResolvedValue({ response: acr });
  });

  test("it returns a handler that removes the agent", async () => {
    handler(webId, policyName);

    expect(setLoading).toHaveBeenCalledWith(true);
    await waitFor(() => {
      expect(accessControl.removeAgentFromPolicy).toHaveBeenCalledWith(
        webId,
        policyName
      );
    });
    expect(mutateResourceInfo).toHaveBeenCalledWith(acr, false);
    expect(setLocalAccess).toHaveBeenCalledWith(null);
  });

  it("handles the public agent", async () => {
    handler(PUBLIC_AGENT_PREDICATE, policyName);

    expect(setLoading).toHaveBeenCalledWith(true);
    await waitFor(() => {
      expect(accessControl.removeAgentFromPolicy).toHaveBeenCalledWith(
        PUBLIC_AGENT_PREDICATE,
        policyName
      );
    });
    expect(accessControl.setRulePublic).toHaveBeenCalledWith(policyName, false);
    expect(mutateResourceInfo).toHaveBeenCalledWith(acr, false);
    expect(setLocalAccess).toHaveBeenCalledWith(null);
  });

  it("handles the authenticated agent", async () => {
    handler(AUTHENTICATED_AGENT_PREDICATE, policyName);

    expect(setLoading).toHaveBeenCalledWith(true);
    await waitFor(() => {
      expect(accessControl.removeAgentFromPolicy).toHaveBeenCalledWith(
        AUTHENTICATED_AGENT_PREDICATE,
        policyName
      );
    });
    expect(accessControl.setRuleAuthenticated).toHaveBeenCalledWith(
      policyName,
      false
    );
    expect(mutateResourceInfo).toHaveBeenCalledWith(acr, false);
    expect(setLocalAccess).toHaveBeenCalledWith(null);
  });
});
