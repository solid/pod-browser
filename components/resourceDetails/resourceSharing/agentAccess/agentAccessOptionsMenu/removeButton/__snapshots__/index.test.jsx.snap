// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentAccessOptionsMenu renders a button which triggers the opening of the menu 1`] = `
<DocumentFragment>
  <div
    aria-disabled="false"
    class="PodBrowser-root PodBrowser-root PodBrowser-gutters PodBrowser-button"
    data-testid="remove-button"
    role="button"
    tabindex="0"
  >
    <div
      class="PodBrowser-root"
    >
      Remove
    </div>
    <span
      class="PodBrowser-root"
    />
  </div>
</DocumentFragment>
`;
